contracts/math/uints.ral

```diff
@@ -90,5 +90,76 @@ Abstract Contract U512Impl () {
                 lower: deltaL
             }
         }
+
+    pub fn bigAdd512(a: U512, b: U512) -> U512 {
+        let (r0, overflow) = overflowingAdd(a.lower, b.lower)
+        let r1 = wrappingAdd(wrappingAdd(a.higher, b.higher), overflow)
+        return U512 { higher: r1, lower: r0 }
+    }
+
+    pub fn bigMul(a: U256, b: U256, denominator: U256) -> U512 {
+        let aLower = low128(a)
+        let aHigher = high128(a)
+        let bLower = low128(b)
+        let bHigher = high128(b)
+
+        let rLowerLower = aLower * bLower
+        let rHigherHigher = aHigher * bHigher
+        let rHigherLower = aHigher * bLower
+        let rLowerHigher = aLower * bHigher
+
+        let mut result = bigAdd512(U512 { higher: rHigherHigher, lower: rLowerLower }, U512 { higher: high128(rHigherLower), lower: low128(rHigherLower) << 128 })
```

Comment:
I would adjust the naming `rHigherHigher` `rLowerLower` does not seem to be the best option